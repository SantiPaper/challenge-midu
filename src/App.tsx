import { StyledContainer } from "./style";
import { Header } from "./components/Header";
import { Main } from "./components/Main";

function App() {
  return (
    <>
      <StyledContainer>
        <Header />
        <Main />
      </StyledContainer>
    </>
  );
}

export default App;
