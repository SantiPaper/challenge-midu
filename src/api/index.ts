import books from "./books.json";
import type { Book } from "../types/books";

export const getBooks = (): Promise<Book[]> => {
  return Promise.resolve(books.library.map((book) => book.book));
};
