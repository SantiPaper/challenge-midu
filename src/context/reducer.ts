import type { Reducer, State } from "../types/context";

export const initialState: State = {
  books: [],
  readingList: [],
};

export const reducer: Reducer = (state = initialState, action) => {
  switch (action.type) {
    case "ADD_BOOKS": {
      return {
        ...state,
        books: action.payload,
      };
    }
    default:
      return state;
  }
};
