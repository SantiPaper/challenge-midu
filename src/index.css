* {
  padding: 0;
  margin: 0;
  box-sizing: border-box;
}

html {
  font-size: 10px;
  --primary-200: #e7cc90;
  --primary-400: #feb510;
  --secondary-400: #788ee5;
  --secondary-200: #9ca9df;
  --black: #090f15;

  --font-xl: 96px;
  --font-l: 64px;
  --font-m: 36px;
  --font-s: 24px;
  --font-xs: 20px;
  --font-xxs: 14px;

  --font-heading: "Yeseva One", cursive;
  --font-body: "Inter", Helvetica, Arial, Sans-serif;
}

body {
  background-color: var(--primary-400);
  font-family: var(--font-body);
}
