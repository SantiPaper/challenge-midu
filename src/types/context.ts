import { Book } from "./books";

export type Context = {
  state: State;
  addBooks: (books: Book[]) => void;
  getByGenre: (genre: Genre) => void;
  getBySearch: (search: string) => void;
  getByPages: (pages: number) => void;
  addToLectureList: (book: Book) => void;
  removeFromLectureList: (book: Book) => void;
  /* mostrarModal: (book: Book) => void; */
  setDetail: (book: Book) => void;
};

export type Genre = "Ciencia ficción" | "Fantasía" | "Terror" | "Zombies" | "";

export type State = {
  books: Book[];
  readingList: Book[];
  filteredBooks: Book[];
  detail: Book[];
};

export type Reducer = (state: State, action: Action) => State;

export type Action =
  | {
      type: "ADD_BOOKS";
      payload: Book[];
    }
  | {
      type: "GET_BY_GENRE";
      payload: Book[];
    }
  | {
      type: "GET_BY_SEARCH";
      payload: Book[];
    }
  | {
      type: "GET_BY_PAGES";
      payload: Book[];
    }
  | {
      type: "ADD_TO_LECTURE_LIST";
      payload: Book;
    }
  | {
      type: "REMOVE_FROM_LECTURE_LIST";
      payload: string;
    }
  | {
      type: "SET_DETAIL";
      payload: Book;
    };
