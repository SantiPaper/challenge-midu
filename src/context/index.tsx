import { createContext, PropsWithChildren, useReducer } from "react";
import type { Context, Genre } from "../types/context";
import { reducer, initialState } from "./reducer";
import { Book } from "../types/books";

export const BooksContext = createContext<Context | null>(null);

export const Provider = ({ children }: PropsWithChildren) => {
  const [state, dispatch] = useReducer(reducer, initialState);

  const addBooks = (books: Book[]) => {
    dispatch({
      type: "ADD_BOOKS",
      payload: books,
    });
  };

  const getByGenre = (genre: Genre) => {
    dispatch({
      type: "GET_BY_GENRE",
      payload: genre
        ? state.books.filter((book) => book.genre === genre)
        : state.books,
    });
  };

  const getBySearch = (search: string) => {
    dispatch({
      type: "GET_BY_SEARCH",
      payload: search
        ? state.books.filter((book) =>
            book.title.toLowerCase().includes(search.toLowerCase())
          )
        : state.books,
    });
  };

  const getByPages = (pages: number) => {
    dispatch({
      type: "GET_BY_SEARCH",
      payload: state.books.filter((book) => book.pages > pages),
    });
  };

  const addToLectureList = (book: Book) => {
    dispatch({
      type: "ADD_TO_LECTURE_LIST",
      payload: book,
    });
  };

  const removeFromLectureList = (book: Book) => {
    dispatch({
      type: "REMOVE_FROM_LECTURE_LIST",
      payload: book.ISBN,
    });
  };

  const setDetail = (book: Book) => {
    dispatch({
      type: "SET_DETAIL",
      payload: book,
    });
  };

  return (
    <BooksContext.Provider
      value={{
        state,
        addBooks,
        getByGenre,
        getBySearch,
        getByPages,
        addToLectureList,
        removeFromLectureList,
        /* showDetail,
        detail,
        mostrarModal, */
        setDetail,
      }}
    >
      {children}
    </BooksContext.Provider>
  );
};
