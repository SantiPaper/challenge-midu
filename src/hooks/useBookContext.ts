import { useContext } from "react";
import { BooksContext } from "../context";

export const useBookContext = () => {
  const value = useContext(BooksContext);
  if (!value) {
    throw new Error("useBookContext tiene que usarse dentro del provider");
  }
  return value;
};
