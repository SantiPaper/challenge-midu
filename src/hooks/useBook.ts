import { useState } from "react";
import { useBookContext } from "./useBookContext";
import type { Book } from "../types/books";

export const useBooks = () => {
  const { setDetail, removeFromLectureList, addToLectureList, state } =
    useBookContext();
  const [showDetail, setShowDetail] = useState(false);

  const closeModal = () => {
    setShowDetail(false);
  };

  const openModal = (book: Book) => {
    setDetail(book);
    setShowDetail(true);
  };

  const onClick = (book: Book) => {
    if (state.readingList.find((b) => b.ISBN === book.ISBN)) {
      removeFromLectureList(book);
    } else addToLectureList(book);
  };

  return { showDetail, openModal, closeModal, onClick };
};
