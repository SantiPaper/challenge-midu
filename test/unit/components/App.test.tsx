import { render } from "@testing-library/react";
import App from "../../../src/App";
import { Provider } from "../../../src/context";
import { mockArr } from "../mock";
import { vi } from "vitest";
import * as api from "../../../src/api";
import * as useBookContext from "../../../src/hooks/useBookContext";
import { Context } from "../../../src/types/context";

describe("Componente App", () => {
  //Otra forma
  /* vi.spyOn(api, "getBooks").mockImplementation(() => {
    return Promise.resolve(mockArr);
  }); */

  vi.spyOn(api, "getBooks").mockResolvedValue(mockArr);
  vi.spyOn(useBookContext, "useBookContext").mockReturnValue({
    addBooks: vi.fn(),
  } as unknown as Context);

  test("Renderiza toda la pagina", () => {
    const { container } = render(
      <Provider>
        <App />
      </Provider>
    );

    expect(container).toMatchSnapshot();
  });
});
