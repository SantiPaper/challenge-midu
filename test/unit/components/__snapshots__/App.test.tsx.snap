// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Componente App > Renderiza toda la pagina 1`] = `
<div>
  <div
    class="sc-bcPKhP evQUrr"
  >
    <header
      class="sc-grXZZQ gYGkhX"
    >
      <nav>
        <ul
          class="header__ul"
        >
          <li>
            <a
              class="header__link"
              href="#"
            >
              Inicio
            </a>
          </li>
          <li>
            <a
              class="header__link"
              href="#"
            >
              Lista de lectura
            </a>
          </li>
        </ul>
      </nav>
    </header>
    <main
      class="sc-dkjKgF"
    >
      <section
        class="sc-hLclGa fgAhUg"
      >
        <h1
          class="hero__title"
        >
          Midubooks
        </h1>
        <p
          class="hero__description"
        >
          Encontrá todos tus libros deseados
        </p>
        <a
          class="hero__button"
          href="#"
        >
          ¡Comenzá ya!
        </a>
      </section>
      <form>
        <label
          for="Búsqueda"
        >
          Buscar libro
        </label>
        <input
          aria-label="Buscar libro por nombre"
          id="Búsqueda"
          type="text"
        />
        <label
          for="Genero"
        >
          Filtrar por genero
        </label>
        <select
          id="Gereno"
          name=""
        >
          <option
            value=""
          >
            Todos
          </option>
          <option
            value="Ciencia ficción"
          >
            Ciencia ficción
          </option>
          <option
            value="Fantasía"
          >
            Fantasía
          </option>
          <option
            value="Terror"
          >
            Terror
          </option>
          <option
            value="Zombies"
          >
            Zombies
          </option>
        </select>
      </form>
      <section
        class="sc-eDnVMP bTuAlC"
      />
    </main>
  </div>
</div>
`;
